[extension/healthcheckv2] Add HTTP service

mwear opened the pull request:
**Description:**

This PR is the third in a series to decompose https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30673 into more manageable pieces for review. This PR introduces the HTTP service which builds upon the aggregation logic added in the [previous PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32695). Following this will be a PR to add a gRPC health check service.

A summary of the changes is below:
- http service based on component status
- supports legacy behavior and config; to be deprecated
- overall collector health can be monitored as well as pipeline health
	- additionally the verbosity of the response can be controlled by passing a `verbose` query parameter
- adds optional endpoint to retrieve the config of the running collector
	- this is currently unredacted JSON and is opt-in

Note, that there will be a follow up PR to add the gRPC service. This will be relevant when reviewing the extension code. It is setup to manage a slice of subcomponents. The HTTP and gRPC services are the subcomponents to be managed, and both implement the `Component` interface. See the [reference PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/30673) for details.

I've provided some examples below to help get an idea of what the responses look like when serialized as JSON.

**Collector Health Example Response**

Below is an example verbose response for the overall collector health. Note, the top level fields correspond to the collector overall, the next level corresponds to the pipelines, and the level below that is the health of the individual components in each pipeline.

```json
{
    "start_time": "2024-01-18T17:27:12.570394-08:00",
    "healthy": true,
    "status": "StatusRecoverableError",
    "error": "rpc error: code = ResourceExhausted desc = resource exhausted",
    "status_time": "2024-01-18T17:27:32.572301-08:00",
    "components": {
        "extensions": {
            "healthy": true,
            "status": "StatusOK",
            "status_time": "2024-01-18T17:27:12.570428-08:00",
            "components": {
                "extension:healthcheckv2": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.570428-08:00"
                }
            }
        },
        "pipeline:metrics/grpc": {
            "healthy": true,
            "status": "StatusRecoverableError",
            "error": "rpc error: code = ResourceExhausted desc = resource exhausted",
            "status_time": "2024-01-18T17:27:32.572301-08:00",
            "components": {
                "exporter:otlp/staging": {
                    "healthy": true,
                    "status": "StatusRecoverableError",
                    "error": "rpc error: code = ResourceExhausted desc = resource exhausted",
                    "status_time": "2024-01-18T17:27:32.572301-08:00"
                },
                "processor:batch": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571132-08:00"
                },
                "receiver:otlp": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571576-08:00"
                }
            }
        },
        "pipeline:traces/http": {
            "healthy": true,
            "status": "StatusOK",
            "status_time": "2024-01-18T17:27:12.571625-08:00",
            "components": {
                "exporter:otlphttp/staging": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571615-08:00"
                },
                "processor:batch": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571621-08:00"
                },
                "receiver:otlp": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571625-08:00"
                }
            }
        }
    }
}
```

**Pipeline Health Example Response**

This is an example verbose response for the `traces/http` pipeline. The top level corresponds to the health of the pipeline, and the second level contains the health of the individual components that make up the pipeline.

```json
{
    "start_time": "2024-01-18T17:27:12.570394-08:00",
    "healthy": true,
    "status": "StatusRecoverableError",
    "error": "rpc error: code = ResourceExhausted desc = resource exhausted",
    "status_time": "2024-01-18T17:27:32.572301-08:00",
    "components": {
        "extensions": {
            "healthy": true,
            "status": "StatusOK",
            "status_time": "2024-01-18T17:27:12.570428-08:00",
            "components": {
                "extension:healthcheckv2": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.570428-08:00"
                }
            }
        },
        "pipeline:metrics/grpc": {
            "healthy": true,
            "status": "StatusRecoverableError",
            "error": "rpc error: code = ResourceExhausted desc = resource exhausted",
            "status_time": "2024-01-18T17:27:32.572301-08:00",
            "components": {
                "exporter:otlp/staging": {
                    "healthy": true,
                    "status": "StatusRecoverableError",
                    "error": "rpc error: code = ResourceExhausted desc = resource exhausted",
                    "status_time": "2024-01-18T17:27:32.572301-08:00"
                },
                "processor:batch": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571132-08:00"
                },
                "receiver:otlp": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571576-08:00"
                }
            }
        },
        "pipeline:traces/http": {
            "healthy": true,
            "status": "StatusOK",
            "status_time": "2024-01-18T17:27:12.571625-08:00",
            "components": {
                "exporter:otlphttp/staging": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571615-08:00"
                },
                "processor:batch": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571621-08:00"
                },
                "receiver:otlp": {
                    "healthy": true,
                    "status": "StatusOK",
                    "status_time": "2024-01-18T17:27:12.571625-08:00"
                }
            }
        }
    }
}
```

**Link to tracking Issue:** #26661

**Testing:** Units / manual

**Documentation:** Comments, etc

mwear committed:
Add HTTP service to healthcheckv2extension

codeboten commented on extension/healthcheckv2extension/internal/http/server_test.go:
```diff
@@ -0,0 +1,4089 @@
+							"pipeline:traces": {
+								healthy: true,
+								status:  component.StatusStopping,
+								nestedStatus: map[string]*componentStatusExpectation{
```
might be worth adding a reusable variable for the common nestedStatus values for each pipelines

codeboten replied:
here's the suggestion for refactoring https://github.com/mwear/opentelemetry-collector-contrib/pull/12294

mwear replied:
thanks for the suggestion. i merged it and found other places the strategy could be applied.

codeboten commented on extension/healthcheckv2extension/internal/http/server_test.go:
```diff
@@ -0,0 +1,4089 @@
+						)
+					},
+					expectedStatusCode: http.StatusOK,
+					expectedComponentStatus: &componentStatusExpectation{
```
same as above, a const for common component status would be great here

codeboten reviewed (approved):
Just a few comments to reduce some duplication in the server test, otherwise i think this looks ok to me

codeboten committed:
refactor test code slightly

Signed-off-by: Alex Boten <223565+codeboten@users.noreply.github.com>

mwear committed:
fixes to suggested refactor

mwear committed:
Continue test simplification

mwear committed:
Remove extension changes to reduce PR size

mwear commented:
@jpkrohling, I removed the extension changes to reduce the PR size. Alex's suggestion helped to shave 1k lines from server_test. The reason it is so large is that there are a number of combinations of settings that are being tested. It tests the response for when permanent error is included or excluded, the same for recoverable error, whether or not to use the legacy or v2 response bodies, and whether or not the body should be verbose. The remainder of the internal/http package will be hard to break up any further. The layout is very modular and you can look at it file by file. Here's a rough roadmap of what you'll find in each:

* server.go
  * configures the handlers: one for the status endpoint and one for the config endpoint
    * either or both can be disabled 
  * determines which responder to use, this will format the response properly (e.g. legacy default response format, legacy custom format, or v2 format)
* handlers.go
  * statusHandler
    * retrieves an aggregate status for the given pipeline and verbosity and passes to a responder that formats and writes the response
  * configHandler
    * renders the active collector config as JSON
* responders.go
  * A responder writes an aggregate status to the http response based on the configured format
  * There are responders for:
    * The new v2 format
    * The legacy formats: the "default" format, or the custom format
* serialization.go
  * Provides serialization for the v2 format.
  * Essentially it provides translation to an intermediary representation of status events with struct tags to be rendered as JSON. The JSON rendering is somewhat complex due to the recursive, nested format.

jpkrohling reviewed (approved):
This looks good, but I'm not a big fan of the tests. I think we might have enough coverage on the "business" of the health check as part of the tests in the `status` package, not sure we need to be doing most of the same checks. A few sanity checks here to test the input/output should be sufficient. I'm not sure I like that they are all table tests as well: the table items are so big that I found myself scrolling back to understand the context again. I _think_ having them as separate test functions would have been easier for me to understand.

That said, I know a lot of effort has been put here, and I think the current state brings progress to the component, so I'm approving the PR.

codeboten commented on extension/healthcheckv2extension/internal/http/responders.go:
```diff
@@ -0,0 +1,147 @@
+	w.WriteHeader(code)
+
+	body, _ := json.Marshal(content)
+	_, _ = w.Write(body)
```
an alternative would be to respond with an error in the json payload (and set an error status code in the response)

codeboten replied:
my suggestion was specifically if there's an error in line 43, not so much 44

mwear replied:
@jpkrohling, are you suggesting I introduce a function such as:

```go
func MustWrite(w http.ResponseWriter, data []byte) {
    _, err := w.Write(data)
    if err != nil {
        panic(err)
    }
}
```

If so, I can do this, but I looked through contrib for prior art, and wasn't able to find anything similar. In fact, most call sites look like this one (e.g. `_, _ = w.Write(body)`. If you have an example of what you'd like to see that would be helpful.

jpkrohling replied:
Yes, something like that. I think I'd still prefer the error to be handled, even if no (current) code path leads to that.

mwear replied:
Wouldn't it be better if we logged the error? A panic seems a little extreme in this case.

jaronoff97 replied:
+1 I think it would be better to log the error and also potentially respond with an error as Alex said above.

From the kube perspective: I think panicking here would crash the extension and crash the pod automatically in kubernetes whereas I think it would be better to handle it explicitly and return a bad response to [let the control plane handle it](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-liveness-http-request).

jaronoff97 replied:
ah i see, in that case I think it would make sense to log and move on. I'm slightly panic-averse

mwear replied:
I made updates to handle both the situations discussed. I updated the code to log errors (at warn level) when writing to the connection. I can change this to panic if you'd prefer (cc: @jpkrohling). For the case where the response body cannot be marshaled to json, an error response body will be generated containing the reason. The response code is preserved, so the health check should succeed either way.

jpkrohling replied:
I agree that a panic is an extreme, but if an error should _never_ happen (and hence, ignored), a panic wouldn't be a problem, as it would never happen :-) If we are afraid of a panic, the error should be handled.

I think the current state of the solution is great, no need to panic :-)

mwear committed:
Add explicit handling for verbose=false

A query param of verbose without a value, or any value other than
false will result in a verbose response body.

mwear commented on extension/healthcheckv2extension/internal/http/handlers.go:
```diff
@@ -0,0 +1,40 @@
+func (s *Server) statusHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		pipeline := r.URL.Query().Get("pipeline")
+		verbose := r.URL.Query().Has("verbose")
```
The intent is to either have the verbose flag or not (without a value), but I can see someone wanting to pass an explicit false. I've updated the code to handle this case. verbose without a value, or any value other than false will result in a verbose response body.

mwear committed:
Log errors writing http response; handle json.Marshal error
